arroyo-state/src/parquet.rs

```diff
@@ -378,48 +438,148 @@ impl BackingStore for ParquetBackend {
                 .storage
                 .get(&file.file)
                 .await
-                .unwrap_or_else(|_| panic!("unable to find file {} in checkpoint", file.file));
-            for (_timestamp, key, value) in
-                self.triples_from_parquet_bytes(bytes.into(), &self.task_info.key_range)
+                .unwrap_or_else(|_| panic!("unable to find file {} in checkpoint", file.file))
+                .into();
+            for (_timestamp, key, value, operation) in
+                self.tuples_from_parquet_bytes(bytes, key_range)
             {
-                state_map.insert(key, value);
+                match operation {
+                    DataOperation::Insert => {
+                        state_map.insert(key, value.unwrap());
+                    }
+                    DataOperation::DeleteKey => {
+                        state_map.remove(&key);
+                    }
+                }
             }
         }
         state_map.into_iter().collect()
     }
+
+    async fn load_compacted(&mut self, compaction: CompactionResult) {
+        self.writer.load_compacted_data(compaction).await;
+    }
 }
 
 impl ParquetBackend {
+    async fn compact_table_partition(
+        table_char: char,
+        task: TaskInfo,
+        epoch_files: BTreeMap<u32, Vec<ParquetStoreData>>,
+        storage_client: StorageProvider,
+        new_min_epoch: u32,
+        table_descriptor: &TableDescriptor,
+    ) -> Option<ParquetStoreData> {
+        // accumulate this partition's tuples from all the table's files
+        // (spread across multiple epochs and files)
+        let mut tuples_in = vec![];
+        for (_epoch, files) in epoch_files {
+            for file in files {
+                // TODO: should we exclude files that are already compact?
```

Comment:
I think probably for the initial version. We'll want phased compaction in the future, but I worry that not doing so means we'll end up copying all live data at whatever the compaction interval is.

Reply:
I've changed this to only compact files of the specified generation, so this shouldn't be a problem.